proposals/handler.go

```diff
@@ -502,22 +508,42 @@ func (h *Handler) checkBallotDataIntegrity(ctx context.Context, b *types.Ballot)
 			epoch-- // download activesets in the previous epoch too
 		}
 		if b.Layer.GetEpoch() >= epoch {
-			if err := h.fetcher.GetActiveSet(ctx, b.EpochData.ActiveSetHash); err != nil {
-				return nil, err
-			}
-			set, err := activesets.Get(h.db, b.EpochData.ActiveSetHash)
-			if err != nil {
-				return nil, err
-			}
-			if len(set.Set) == 0 {
-				return nil, fmt.Errorf("%w: empty active set ballot %s", pubsub.ErrValidationReject, b.ID().String())
+			var exists bool
+			totalWeight, exists := h.activeSets.Get(b.EpochData.ActiveSetHash)
+			if !exists {
+				if err := h.fetcher.GetActiveSet(ctx, b.EpochData.ActiveSetHash); err != nil {
+					return nil, err
+				}
+				set, err := activesets.Get(h.db, b.EpochData.ActiveSetHash)
+				if err != nil {
+					return nil, err
+				}
+				if len(set.Set) == 0 {
+					return nil, fmt.Errorf(
+						"%w: empty active set ballot %s",
+						pubsub.ErrValidationReject,
+						b.ID().String(),
+					)
+				}
+				computed, used := h.atxsdata.WeightForSet(set.Epoch, set.Set)
+				for i := range used {
+					if !used[i] {
+						return nil, fmt.Errorf(
+							"missing atx %s in active set ballot %s",
+							set.Set[i].ShortString(),
+							b.ID().String(),
+						)
+					}
+				}
+				totalWeight = computed
+				h.activeSets.Add(b.EpochData.ActiveSetHash, totalWeight)
 			}
-			actives = set.Set
+			return &RefBallotAux{TotalWeight: totalWeight}, nil
 		}
 	} else if b.EpochData != nil {
 		return nil, errUnexpectedEpochData
 	}
-	return actives, nil
+	return nil, nil
```

Comment:
To be compatible with the previous version of the code this can maybe return

```suggestion
	return &RefBallotAux{TotalWeight: 0}, nil
```

because before it would return a slice of length 0

Reply:
refactored